pkg/cmd/roachtest/test_runner.go

```diff
@@ -1273,24 +1318,48 @@ func (r *testRunner) teardownTest(
 	ctx context.Context, t *testImpl, c *clusterImpl, timedOut bool,
 ) error {
 	var err error
+
 	if timedOut || t.Failed() {
 		err = r.collectArtifacts(ctx, t, c, timedOut, time.Hour)
 		if err != nil {
 			t.L().Printf("error collecting artifacts: %v", err)
 		}
-	}
 
-	if timedOut {
-		// Shut down the cluster. We only do this on timeout to help the test terminate;
-		// for regular failures, if the --debug flag is used, we want the cluster to stay
-		// around so someone can poke at it.
-		_ = c.StopE(ctx, t.L(), option.DefaultStopOpts(), c.All())
-
-		// We previously added a timeout failure without cancellation, so we cancel here.
-		if t.mu.cancel != nil {
-			t.mu.cancel()
+		if timedOut {
+			// Shut down the cluster. We only do this on timeout to help the test terminate;
+			// for regular failures, if the --debug flag is used, we want the cluster to stay
+			// around so someone can poke at it.
+			_ = c.StopE(ctx, t.L(), option.DefaultStopOpts(), c.All())
+
+			// We previously added a timeout failure without cancellation, so we cancel here.
+			if t.mu.cancel != nil {
+				t.mu.cancel()
+			}
+			t.L().Printf("test timed out; check __stacks.log and CRDB logs for goroutine dumps")
 		}
-		t.L().Printf("test timed out; check __stacks.log and CRDB logs for goroutine dumps")
+	} else if t.goCoverEnabled {
+		// Go cover data is dumped when the cockroach process exits (it is not
+		// dumped if the process is killed). We want the process to terminate as
+		// soon as possible. We send two SIGINTs; the first one starts draining and
+		// the second causes the process to exit right away.
+		//
+		// TODO(radu): many tests stop the nodes with the default options
+		// (SIGKILL) during the test, which means some coverage data will be lost.
+		t.L().Printf("Shutting down cluster gracefully to obtain go cover artifacts")
+		stopOpts := option.DefaultStopOpts()
+		stopOpts.RoachprodOpts.Sig = 2 // SIGINT
+		stopOpts.RoachprodOpts.Wait = false
+		if err := c.StopE(ctx, t.L(), stopOpts, c.All()); err != nil {
+			t.L().PrintfCtx(ctx, "error stopping cluster: %v", err)
+		}
+		stopOpts.RoachprodOpts.Sig = 2 // SIGINT
+		stopOpts.RoachprodOpts.Wait = true
+		stopOpts.RoachprodOpts.MaxWait = 30
+		if err := c.StopE(ctx, t.L(), stopOpts, c.All()); err != nil {
+			t.L().PrintfCtx(ctx, "error stopping cluster: %v", err)
+		}
+		t.L().Printf("Retrieving go cover artifacts")
+		getGoCoverArtifacts(ctx, t.L(), c, t)
```

Comment:
I don't get it. Why are we only fetching coverage data when the test failed for non-timeout reasons? Aren't we interested in coverage data for all runs?